### PR TITLE
Feature: Multiple cursors

### PR DESCRIPTION
**Pull request by rvilarl**
This feature allows to configure multiple cursors with OSMDOptions.

### CODE REVIEW

**Comment by rvilarl** [comment]
I leave a note to make the review easier. Basicly the change consist on:
- A new element in the OSMDOptions.ts which is an array of CursorOptions defining the cursors. The parameters are documented in the file. If not defined, one standard cursor is defined in the method SetOptions of OpenSheetMusicDisplay.
- Changes in OpenSheetMusicDisplay.ts to allow multiple cursors. Main difference from an user perspective is that the "cursor" public variable is now "cursors" an array of cursors.
- Changes on the test file to reference cursors[0] instead of cursor.
- Cursor.ts constructor and UpdateStyle accept now an instance of CursorOptions. Code to implement the 4 types of cursors:
     0: Standard highlighting current notes
     1: Thin line left to the current notes
     2: Short thin line on top of stave and left to the current notes
     3: Current compass
     4: Current compass to left of current notes

**Comment by rvilarl** [comment]
Some examples of the five different types of cursors , they are not at the same score location

**TYPE 2(grey) & TYPE 1 (green)**
in this example showing metronome location (grey) and left of the next notes to play(green) like in https://michaelecke.com/pianoplay/
image


**TYPE 0**
image

**TYPE 1**
image

**TYPE 2**
image

**TYPE 3**
image

**TYPE 4**
image

**Comment by rvilarl** [comment]
@sschmidTU is there anything that you would like to get changed? Do you still believe that this is a good idea?

**Comment by sschmidTU** [comment]
i just didn't get to review it yet, don't worry.

**Comment by rvilarl** [comment]
@sschmidTU do you want me to resolve the conflicts?

**Comment by rvilarl** [comment]
I did it just in case

**Comment by conde2** [comment]
@rvilarl would be cool to add an option to center the cursor, check this issue:

https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1011

**Comment by BppleMan** [comment]
Looking forward to a release containing this PR,I need multiple cursor

File: src/OpenSheetMusicDisplay/OSMDOptions.ts [hunk 1]
```diff
@@ -293,3 +297,27 @@ export interface AutoBeamOptions {
...
+export interface CursorOptions {
+    /**
+     * Type of cursor:
+     * 0: Standard highlighting current notes
+     * 1: Thin line left to the current notes
+     * 2: Short thin line on top of stave and left to the current notes
+     * 3: Current compass
+     * 4: Current compass to left of current notes
+     */
+    type: number;
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OSMDOptions.ts:310]: I would prefer an enum here with named members. You can still do it like this and use numbers:
```ts
export enum CursorType {
  Standard: 0,
  ThinLeft: 1,
  ShortThinTopLeft: 2,
  CurrentArea,
  CurrentAreaLeft
}
```

If you find better names for these, that's fine.
By the way, what does "compass" mean? That has absolutely no meaning for me in the context of music. Is it the measure? The currently playing notes? I'd really like a more obvious name here.

**Comment by sschmidTU** [comment]
This will probably be merged for the next release.

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts [hunk 2]
```diff
@@ -570,6 +575,12 @@ export class OpenSheetMusicDisplay {
         if (options.autoGenerateMutipleRestMeasuresFromRestMeasures !== undefined) {
             this.rules.AutoGenerateMutipleRestMeasuresFromRestMeasures = options.autoGenerateMutipleRestMeasuresFromRestMeasures;
         }
+        if (options.cursorsOptions !== undefined)
+        {
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:579]: please adjust the bracket style to what we use, bracket on same line in if blocks.

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts [hunk 3]
```diff
@@ -740,41 +753,42 @@ export class OpenSheetMusicDisplay {
...
-                backendToDrawOn = this.drawer.Backends[newPageNumber - 1];
-            }
-            // create new cursor
-            if (backendToDrawOn && backendToDrawOn.getRenderElement()) {
-                this.cursor = new Cursor(backendToDrawOn.getRenderElement(), this);
-            }
-            if (this.sheet && this.graphic && this.cursor) { // else init is called in load()
-                this.cursor.init(this.sheet.MusicPartManager, this.graphic);
-            }
+            for (let i: number = 0; i< this.cursorsOptions.length; i++){
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:759]: space between "i<". shouldn't the linter catch this?

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts [hunk 4]
```diff
@@ -880,11 +894,11 @@ export class OpenSheetMusicDisplay {
     public setDrawBoundingBox(value: string, render: boolean = false): void {
         this.drawBoundingBox = value;
         if (this.drawer) {
-            this.drawer.drawableBoundingBoxElement = value; // drawer is sometimes created anew, losing this value, so it's saved in OSMD now.
+        this.drawer.drawableBoundingBoxElement = value; // drawer is sometimes created anew, losing this value, so it's saved in OSMD now.
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:897]: fix indentation. your IDE or linter should catch this style issue.

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts [hunk 5]
```diff
@@ -880,11 +894,11 @@ export class OpenSheetMusicDisplay {
...
         }
         if (render) {
-            this.render(); // may create new Drawer.
-        }
+        this.render(); // may create new Drawer.
+    }
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:901]: fix indentation.

File: src/OpenSheetMusicDisplay/OSMDOptions.ts [hunk 6]
```diff
@@ -239,6 +239,10 @@ export interface IOSMDOptions {
      * This works across instruments- If all instruments have subsequent measures with nothing but rests, multirest measures are generated
      */
     autoGenerateMutipleRestMeasuresFromRestMeasures?: boolean;
+    cursorsOptions?: CursorOptions[];
+    /**
+     * Defines multiple simultaneous cursors. If left undefined the standard cursor will be used.
+     */
```
Comment (sschmidTU) [inline src/OpenSheetMusicDisplay/OSMDOptions.ts:245]: The docstring needs to be above the variable, not below.

**Review by sschmidTU — CHANGES_REQUESTED** [review]
Looks very good in general, but please adjust the changes requested, which is mostly style, and changing a parameter from number to enum.

**Comment by rvilarl** [comment]
@sschmidTU I will look tomorrow into that, thanks for the feedback!

File: src/OpenSheetMusicDisplay/OSMDOptions.ts (hunk 1 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OSMDOptions.ts:310]: Name are find. You are right I refered to measure, 'compas' is spanish :). Also changed.

Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OSMDOptions.ts:310]: done

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts (hunk 2 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:579]: done

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts (hunk 4 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:897]: my IDE Visual Studio Code applies the eslint configuration in the project. It seems that this is not configured. Done

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts (hunk 5 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:901]: ok

File: src/OpenSheetMusicDisplay/OSMDOptions.ts (hunk 6 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OSMDOptions.ts:245]: done

**Comment by rvilarl** [comment]
I decided to do it right now. Ready!

File: src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts (hunk 3 above)
Comment (rvilarl) [reply src/OpenSheetMusicDisplay/OpenSheetMusicDisplay.ts:758]: done also

**Comment by rvilarl** [comment]
@sschmidTU i forgot one comment, now ready. Will the new release handle repetitions?

**Comment by sschmidTU** [comment]
> Will the new release handle repetitions?

Probably not yet, cursor repetition handling is part of our playback code, which is in the OSMD audio player, which is currently Github Sponsors-exclusive (early access).

**Review by sschmidTU — APPROVED** [review]
Looks good now, thanks! Merging.
Merging to feature branch temporarily, because the index.js needs fixes (still using osmd.cursor), and i'll add a osmd.Cursor getter which returns osmd.cursors[0], as many people will probably only use one cursor regularly.

**Comment by sschmidTU** [comment]
Merged to develop now with above addition.

**Comment by sschmidTU** [comment]
For a usage example see https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/1005#issuecomment-833981960.

**Comment by gerardoscoppetta** [comment]
Well done. This feature is great for projects for studying music notation. All variations count the notes of the current measure. I have a suggestion that would be essential for projects where music reading is studied. I would add another variant of cursor, (5) where, based on the nominator and denominator of the time I would divide the progress of the cursor. For example, if the tempo of those measures is 4/4, I would advance the cursor by 4 beats until the measure is filled, if by 2/2 I would advance the cursor by half a measure by 2. s and by 6/8. I would divide the measurement into six segments and so on. I hope I have expressed the concept well.
Thanks. Hope to find this feature soon.

**Comment by sschmidTU** [comment]
@gerardoscoppetta I'm afraid I'm not sure what this would look like, do you mean basically highlighting a part of a measure (with a "box"), instead of the full measure for example?
If you have some programming experience, why not try to implement it?

**Comment by gerardoscoppetta** [comment]
Hi Simon. I am mainly a musician and a music teacher, with programming I am not an expert but I manage a bit, I can try. Basically when studying music the metronome is used a lot to divide the measures into smaller regular parts. Now if a piece has all measures 3/4 as tempo, if there are no changes in tempo, they last 3 quarter beats for each measure and advance smoothly. My idea is this, visually underline by dividing each measure by the tempo. So, in practice, knowing the width of each measure divides the measure by the metric nominator. For example, if the tempo is 3/4 and the measure 1ed is 90px wide it divides 90/3, if the measure 2 is 120px wide it divides 120/3. and so on. these measures will then be used to under-highlight the progress of the score as if it were a visual metronome.

**Comment by sschmidTU** [comment]
Hey Gerardo, cool, I'm also a musician ;) I play the piano and viola/violin. Nice to see some more music teachers here as well!

I understand what you mean musically, but i'm not so sure what it should look like.
So, you mean something like this?
image
(I just quickly hacked this for this cursor position only, not a general function yet)

Yeah, I can see how that could be useful.

(if that's not what you meant, please try drawing an image, because I'm not sure what your description means for how it looks)

**Comment by gerardoscoppetta** [comment]
Ok Simon. I'll make you a little demonstration video. Time to do it. :)

**Comment by gerardoscoppetta** [comment]
I'm a Pianist too :). Also in order of competence I play drums, guitar and harmonica. I have a small recording studio in Turin.
Here is a little video that you will surely understand what I mean. Basically a visual metronome in the score :)
https://drive.google.com/file/d/1bOMoTWKYmAIMPubEBYLSEW--9nnqnp6g/view?usp=sharing

**Comment by rvilarl** [comment]
Oh I see, you want to have the metronome per quarter note and not per note as in my app or per measure. This would be indeed a new cursor type. In your video though, the first metronome cursor is covering two quarters, is that a failure?

**Comment by gerardoscoppetta** [comment]
Hi Rodrigo. The visual metronome should have the equivalent of the nominator for each measure. For example, if a piece is made up of 4 3/4 measures. Each measure should have the passage of 3 cursors. in time, regardless of how many notes are contained in each measure. more or less as I show in the video.

**Pull request merged**